agent/config/builder.go

```diff
@@ -185,11 +185,18 @@ func (b *Builder) ReadPath(path string) ([]Source, error) {
 			continue
 		}
 
-		src, err := b.ReadFile(fp)
-		if err != nil {
-			return nil, err
+		configFormat := b.stringVal(b.Flags.ConfigFormat)
+		srcFormat := FormatFrom(fp)
+
+		// If config-format is not set, only read files with supported extensions
+		// If config-format is set, files of any extension will be interpreted in that format
+		if srcFormat == "json" || srcFormat == "hcl" || configFormat != "" {
+			src, err := b.ReadFile(fp)
+			if err != nil {
+				return nil, err
+			}
+			sources = append(sources, src)
```

Comment:
Can we add a test to the table test of exceptions to validate this behaviour? The table test in runtime_test.go should make it easy to test I think this as you can provide command options and make assertions on the result and any warnings or errors printed.

Reply:
For reference this is more or less copy-pasted from https://github.com/hashicorp/consul/blob/15e80e4e76a7f207c03da11f32bf8d948ac699cb/agent/config/builder.go#L238-L255 My understanding is that the bug is that we only respect the format flag in the single file case currently and not in the directory case which this is fixing.

I wonder if this is essentially copy-paste whether it would be possible to abstract this into a method like `shouldParseFile`?

Reply:
For tests see https://github.com/hashicorp/consul/blob/15e80e4e76a7f207c03da11f32bf8d948ac699cb/agent/config/runtime_test.go#L385-L399